docker/Dockerfile

```diff
@@ -30,31 +32,31 @@ RUN apt-get update && apt-get install -y \
     libxslt1-dev \
     rustc \
     zlib1g-dev \
- && rm -rf /var/lib/apt/lists/*
+    && rm -rf /var/lib/apt/lists/*
 
 # Build dependencies that are not available as wheels, to speed up rebuilds
 RUN pip install --prefix="/install" --no-warn-script-location \
-        cryptography \
-        frozendict \
-        jaeger-client \
-        opentracing \
-        # Match the version constraints of Synapse
-        "prometheus_client>=0.4.0" \
-        psycopg2 \
-        pycparser \
-        pyrsistent \
-        pyyaml \
-        simplejson \
-        threadloop \
-        thrift
+    cryptography \
+    frozendict \
+    jaeger-client \
+    opentracing \
+    # Match the version constraints of Synapse
+    "prometheus_client>=0.4.0" \
+    psycopg2 \
+    pycparser \
+    pyrsistent \
+    pyyaml \
+    simplejson \
+    threadloop \
+    thrift
 
 # now install synapse and all of the python deps to /install.
 COPY synapse /synapse/synapse/
 COPY scripts /synapse/scripts/
 COPY MANIFEST.in README.rst setup.py synctl /synapse/
 
-RUN pip install --prefix="/install" --no-warn-script-location \
-        /synapse[all]
+RUN --mount=type=cache,mode=0755,target=/root/.cache/pip pip install --prefix="/install" --no-warn-script-location \
```

Comment:
The cache doesn't seem to work. Still takes 30 seconds for this layer.

The BuildKite caches are supposed to work outside of the layer caching so I don't think the invalidated `COPY synapse` should be causing problems with this approach.

Based it off of:

 - https://stackoverflow.com/a/58021389/796832
 - https://gist.github.com/RainJayTsai/16979ee1687546f93827ba78e5657196